File: 백준/Platinum/2150. Strongly Connected Component/Strongly Connected Component.cc
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int scc[10010];
vector<vector<int>> graph;
vector<vector<int>> grapht;
vector<bool> visited;
stack<int> p;

void dfs(int x) {
    visited[x] = true;
    for (int next: graph[x]) {
        if (!visited[next])
            dfs(next);
    }
    p.push(x);
}

void dfs2(int k, int x) {

    visited[x] = true;
    scc[x] = k;
    for (int next: grapht[x]) {
        if (!visited[next])
            dfs2(k, next);
    }
}

int main() {
    int v, e;
    int s = 0;
    cin >> v >> e;
    graph.assign(v + 1, vector<int>(0, 0));
    grapht.assign(v + 1, vector<int>(0, 0));
    visited.assign(v + 1, false);
    for (int i = 0; i < e; i++) {
        int start, end;
        cin >> start >> end;
        graph[start].emplace_back(end);
        grapht[end].emplace_back(start);
    }

    for (int i = 1; i <= v; i++) {
        if (!visited[i])
            dfs(i);
    }

    visited.assign(v + 1, false);

    while (!p.empty()) {
        int c = p.top();
        p.pop();
        if (visited[c])
            continue;

        dfs2(++s, c);
    }

    cout << s << endl;
    for (int i = 1; i <= v; i++) {
        if (scc[i] == -1)
            continue;
        cout << i << " ";
        for (int j = i + 1; j <= v; j++) {
            if (scc[j] == scc[i]) {
                cout << j << " ";
                scc[j] = -1;
            }
        }
        cout << -1 << endl;
    }
    return 0;
}

File: 백준/Platinum/2150. Strongly Connected Component/Strongly Connected Component.py
import sys

sys.setrecursionlimit(10**6)

scc = [0] * 10010
p = []

v, e = map(int, input().split())
s = 0

graph = {x: [] for x in range(v + 1)}
grapht = {x: [] for x in range(v + 1)}
visited = [False for _ in range(v + 1)]


def dfs(x):
    visited[x] = True
    for node in graph[x]:
        if not visited[node]:
            dfs(node)
    p.append(x)


def dfs2(k, x):
    visited[x] = True
    scc[x] = k
    for node in grapht[x]:
        if not visited[node]:
            dfs2(k, node)


for i in range(e):
    start, end = map(int, input().split())
    graph[start].append(end)
    grapht[end].append(start)

for i in range(1, v + 1):
    if not visited[i]:
        dfs(i)

visited = [False] * (v + 1)

while p:
    c = p.pop()
    if visited[c]:
        continue
    s += 1
    dfs2(s, c)

print(s)
for i in range(1, v + 1):
    if scc[i] == -1:
        continue
    print(i, end=" ")
    for j in range(i + 1, v + 1):
        if scc[j] == scc[i]:
            print(j, end=" ")
            scc[j] = -1
    print(-1)
